ResponseTimer callstack exceeded

dimitropoulos opened the pull request:
closes INS-698

dimitropoulos committed:
no default export (prepping for hooks)

dimitropoulos committed:
makes logic match other implementation

note that above in `componentDidUpdate` it uses `<= 0`, whereas here it uses `> 0` but then only checks for false.

Also, logically, there's no way to ever get `aria-hidden="true"` since it returns early so that entire attribute can just be removed.

dimitropoulos committed:
removes fake private class member syntax

this is moving to hooks anyway, but in the meantime...

dimitropoulos committed:
undoes calling identical code twice in a row...

dimitropoulos committed:
condenses class methods to prepare for hooks

dimitropoulos committed:
hooks refactor

fairly faithful to the original, this attempts to just refactor to hooks.  this exposes, again, the fact that we're setting state within useEffect.

dimitropoulos committed:
removes 200 ms offset for response time

so that, now, you know, the time reported is the actual time.

dimitropoulos committed:
adds logging for response timer lage

develohpanda commented on packages/insomnia-app/app/ui/components/response-timer.tsx:
```diff
@@ -1,84 +1,61 @@
-      return;
-    }
+    };
+  }, [loadStartTime, setMilliseconds, isLoading, responseTime]);
```
It certainly feels strange for `responseTime` to be a dependency for _this_ effect, the intent of which is simply to increment `milliseconds`, right? It can be broken up into an individual effect to handle the lag, and a custom hook/individual effect to handle the timer incrementing. The custom hook can then, also, be unit tested separately.

Here is an example implementation: this is largely the same as your implementation but separated into standalone effects.
* `milliseconds` is reset each time the timer starts (not when it clears)
* the timing logic is extracted to a custom hook and standalone effect
* calculation of the lag is handled by a standalone effect

```ts
import React, { DOMAttributes, FunctionComponent, useEffect, useState } from 'react';
import { REQUEST_TIME_TO_SHOW_COUNTER } from '../../common/constants';

interface Props {
  handleCancel: DOMAttributes<HTMLButtonElement>['onClick'];
  loadStartTime: number;
  responseTime?: number;
}

const useTimer = (startTime: number, intervalMs: number) => {
  const [milliseconds, setMilliseconds] = useState(0);
  const [running, setRunning] = useState(false);

  useEffect(() => {
    let interval: NodeJS.Timeout | null = null;
    if (startTime > 0) {
      setMilliseconds(0);
      setRunning(true);
      interval = setInterval(() => {
        const milliseconds = Date.now() - startTime;
        setMilliseconds(milliseconds);
      }, intervalMs);
    }
    return () => {
      if (interval !== null) {
        clearInterval(interval);
        setRunning(false);
      }
    };
  }, [startTime, intervalMs]);

  return { milliseconds, running };
};

export const ResponseTimer: FunctionComponent<Props> = ({ handleCancel, loadStartTime, responseTime = 0 }) => {
  const { milliseconds, running } = useTimer(loadStartTime, 100);

  // When we have a response time, and an elapsed time, and have stopped running, log the lag
  useEffect(() => {
    if (!running && milliseconds > 0 && responseTime > 0) {
      console.info(`[ResponseTimer] request setup and tear-down measured to be ${Math.floor(milliseconds - responseTime)}ms`, {
        totalDuration: milliseconds,
        networkDuration: responseTime,
      });
    }
  }, [running, milliseconds, responseTime]);

  if (!running) {
    return null;
  }

  const seconds = milliseconds / 1000;
  return (
    <div className="overlay theme--transparent-overlay">
      <h2>
        {seconds >= REQUEST_TIME_TO_SHOW_COUNTER ? `${seconds.toFixed(1)} seconds` : 'Loading'}...
      </h2>
      <div className="pad">
        <i className="fa fa-refresh fa-spin" />
      </div>
      <div className="pad">
        <button
          className="btn btn--clicky"
          onClick={handleCancel}
        >
          Cancel Request
        </button>
      </div>
    </div>
  );
};

```

dimitropoulos replied:
I don't see what this accomplishes, to be honest, because the thing you said you were trying to avoid (responseTime being a dependency) was not avoided.  The difference now is just that we introduced new state (i.e. `running`) to share state.

This implementation has a bug, which is that it will log twice when you switch to a new request.  To reproduce:
- make `Request 1`, see the log occur
- switch to `Request 2` and see the log occur again

The trouble is, that _some_ of the state you see logged again is from `Request 1`, e.g. `totalDuration`, while the other is from whatever request you just selected, e.g. `networkDuration`.

-----

All the same, I really see the value here in what you're aiming for and I like the "single responsibility principle" adherence a lot.

I'd like to develop this more because I think there's value in it, but for now because I don't want this PR to stick around for another day, I've just removed the log line altogether.  We really are gonna need a nice `useTimer` bulletproof implementation that fills our need because we have timers like this ALL OVER the app and I'd love to work with you more on developing one that's clean and works well (and has tests), but this bug has reminded me that it's out of scope for the callstack exceeded bug this PR set out to fix.

Thanks, looking forward to making progress on this in the near future.

dimitropoulos replied:
removed in https://github.com/Kong/insomnia/pull/3386/commits/a2e2ddfe46fd8a59358a8d274cefbe0eb9511393

develohpanda replied:
> I don't see what this accomplishes, to be honest, because the thing you said you were trying to avoid (responseTime being a dependency) was not avoided. 

You misunderstood 🙂 there were two effects in the previous implementation bundled into one, and the thing I said was `responseTime` doesn't make sense as a dependency for _this_ effect. The two effects are:
- increment time at a regular interval until unmount (`useTimer()`)
- log the lag once we have a response (standalone effect)

The additional `running` state is to avoid placing the timeout in a shared location.

> This implementation has a bug, which is that it will log twice when you switch to a new request.

Not a rigorously tested code snippet, it was merely an example (as stated) to demonstrate the separation of effects.

However you captured the essence with the `single responsibility principle` - when I was testing the initial implementation I noticed the timer effect was being called multiple times as the `responseTime` prop changed which didn't conceptually make sense - _when this component mounts, start a timer to and show the elapsed time_.

Likely the bug you described above is a conditional I was missing in the "handle lag" effect, but ultimately the two _are_ separate.

I'm glad you've backtracked on handling the lag, though, because that also was also out of scope for the callstack exceeded bug. 👍🏽

develohpanda commented on packages/insomnia-app/app/ui/components/response-timer.tsx:
```diff
@@ -1,84 +1,61 @@
+            console.info(`[ResponseTimer] lag measured to be ${Math.floor(milliseconds - responseTime)}ms`, {
+              wallTime: milliseconds,
+              responseTime,
+            });
```
```suggestion
            console.info(`[ResponseTimer] request setup and tear-down measured to be ${Math.floor(milliseconds - responseTime)}ms`, {
              totalDuration: milliseconds,
              networkDuration: responseTime,
            });
```

dimitropoulos replied:
see https://github.com/Kong/insomnia/pull/3386#discussion_r633498495

develohpanda commented on packages/insomnia-app/app/ui/components/response-timer.tsx:
```diff
@@ -1,84 +1,61 @@
-        clearInterval(this._interval);
+    return () => {
+      if (interval !== null) {
+        setMilliseconds(milliseconds => {
```
Even though this is in it's own closure having two variables with the exact same name (`milliseconds`) in nested code-blocks can be very misleading when skimming through the code.

dimitropoulos replied:
I don't strongly agree, haha, but I guess I wouldn't because I wrote the code. :stuck_out_tongue: 


I see where you're coming from, though, and I removed a similar case in 9506e39f82a41ed5fc31c28e53e48923475e8c16

This line you commented on was removed in https://github.com/Kong/insomnia/pull/3386/commits/a2e2ddfe46fd8a59358a8d274cefbe0eb9511393

develohpanda reviewed (changes requested):
The discrepancy in the times seen could also be a result of this: https://github.com/Kong/insomnia/blob/1271f5a925c956452165fa28384c6dfb84f6fdec/packages/insomnia-app/app/network/network.ts#L927-L947


Also something here doesn't quite add up: 
<img width="1360" alt="image" src="https://user-images.githubusercontent.com/4312346/118432919-f5316b00-b72d-11eb-8f00-b896984f4567.png">

Anyway, a couple of suggested improvements.

dimitropoulos committed:
removes timer logging and `responseTime` prop after PR discussion

it's definitely fruitful, but the fix for the callstack exceeded is what needs to be the focus.

dimitropoulos committed:
removes shadowed variable per PR feedback

dimitropoulos committed:
reinstates and documents 200ms compensation

dimitropoulos commented:
@develohpanda great find on the discrepancy between timers.  Looks like the rabbit hole goes a bit further down.  cool!

And I really liked where you were headed with the `useTimer` implementation.  We're really gonna need something like that in the future, but I think it is a little too much for what I want to accomplish with this PR, which is just to fix the callstack-exceeded bug.

We need to discuss more (and, haha, pun intended, take our _time_ in doing so), have tests, etc. for a `useTimer` implementation that works for us because we do similar things to this with timers all over the app and I think we'll need something like it.  And once we have something we have tested and works well we can contribute it to `react-use`.  I removed this stuff for now but I'm excited to carry forward where we got with this PR.

develohpanda committed:
add delay endpoint to example server

develohpanda commented on packages/insomnia-app/app/ui/components/response-timer.tsx:
```diff
@@ -1,84 +1,51 @@
-      return;
-    }
+    };
+  }, [loadStartTime, setMilliseconds, isLoading]);
```
I believe the setState function isn't needed as a dependency because React ensures it will never change (ES lint doesn't complain either, I'll link docs when I find em)

dimitropoulos replied:
ahh, right right, yes, this is mistake.  it looked a little different for a bit when I was working with it and at the time it was require.  definitely not needed:
> React guarantees that setState function identity is stable and won’t change on re-renders. This is why it’s safe to omit from the useEffect or useCallback dependency list.
> - https://reactjs.org/docs/hooks-reference.html#usestate

develohpanda reviewed (approved):
LGTM, I validated and pushed the endpoint I added in order to test this (which can be used for smoke tests later)

dimitropoulos commented on packages/insomnia-app/app/ui/components/response-timer.tsx:
```diff
@@ -1,84 +1,51 @@
-      return;
-    }
+    };
+  }, [loadStartTime, setMilliseconds, isLoading]);
```
```suggestion
  }, [loadStartTime, isLoading]);
```

dimitropoulos committed:
Update packages/insomnia-app/app/ui/components/response-timer.tsx
